File: App/app/src/main/java/com/cinema/ticket_booking/network/ApiService.java
package com.cinema.ticket_booking.network;

import com.cinema.ticket_booking.data.model.request.*;
import com.cinema.ticket_booking.data.model.response.*;

import java.util.List;
import java.util.Map;

import retrofit2.Call;
import retrofit2.http.*;
import okhttp3.MultipartBody;

public interface ApiService {

        // ── Auth ─────────────────────────────────────────────────────────────
        @POST("auth/register")
        Call<ApiResponse<AuthResponse>> register(@Body RegisterRequest request);

        @POST("auth/login")
        Call<ApiResponse<AuthResponse>> login(@Body LoginRequest request);

        @POST("auth/social-login")
        Call<ApiResponse<AuthResponse>> socialLogin(@Body SocialLoginRequest request);

        @POST("auth/refresh")
        Call<ApiResponse<AuthResponse>> refreshToken(@Body RefreshTokenRequest request);

        @POST("auth/logout")
        Call<ApiResponse<Void>> logout(@Body RefreshTokenRequest request);

        @POST("auth/forgot-password")
        Call<ApiResponse<Void>> forgotPassword(@Body ForgotPasswordRequest request);

        @POST("auth/verify-otp")
        Call<ApiResponse<String>> verifyOtp(@Body Map<String, String> body);

        @POST("auth/reset-password")
        Call<ApiResponse<Void>> resetPassword(@Body ResetPasswordRequest request);

        // ── User ─────────────────────────────────────────────────────────────
        @GET("users/me")
        Call<ApiResponse<UserResponse>> getMyProfile();

        @PATCH("users/me")
        Call<ApiResponse<UserResponse>> updateProfile(@Body UpdateProfileRequest request);

        @PATCH("users/me/notifications")
        Call<ApiResponse<UserResponse>> updateNotificationSettings(@Body NotificationSettingsRequest request);

        @Multipart
        @POST("users/me/avatar")
        Call<ApiResponse<UserResponse>> uploadAvatar(@Part MultipartBody.Part file);

        @PUT("users/me/password")
        Call<ApiResponse<Void>> changePassword(@Body Map<String, String> body);

        @PATCH("users/me/fcm-token")
        Call<ApiResponse<Void>> updateFcmToken(@Query("token") String token);

        // ── Movie ─────────────────────────────────────────────────────────────
        @GET("movies")
        Call<ApiResponse<PageResponse<MovieSummary>>> getMovies(
                        @Query("status") String status,
                        @Query("page") int page,
                        @Query("size") int size);

        @GET("movies/search")
        Call<ApiResponse<PageResponse<MovieSummary>>> searchMovies(
                        @Query("q") String query,
                        @Query("page") int page,
                        @Query("size") int size);

        @GET("movies/{id}")
        Call<ApiResponse<MovieDetail>> getMovieDetail(@Path("id") String id);

        @GET("movies/{id}/can-review")
        Call<ApiResponse<CanReviewResponse>> canReview(@Path("id") String id);

        @GET("movies/cinema/{cinemaId}")
        Call<ApiResponse<List<MovieSummary>>> getNowShowingByCinema(@Path("cinemaId") String cinemaId);

        @GET("movies/genres")
        Call<ApiResponse<List<Genre>>> getAllGenres();

        // ── Cinema ────────────────────────────────────────────────────────────
        @GET("cinemas")
        Call<ApiResponse<List<CinemaResponse>>> getCinemas(@Query("city") String city);

        @GET("cinemas/{id}")
        Call<ApiResponse<CinemaResponse>> getCinemaDetail(@Path("id") String id);

        // ── Showtime ──────────────────────────────────────────────────────────
        @GET("showtimes")
        Call<ApiResponse<List<ShowtimeResponse>>> getShowtimes(
                        @Query("movieId") String movieId,
                        @Query("cinemaId") String cinemaId,
                        @Query("date") String date); // format: yyyy-MM-dd

        @GET("showtimes/{id}/seats")
        Call<ApiResponse<SeatMapResponse>> getSeatMap(@Path("id") String showtimeId);

        // ── Booking ───────────────────────────────────────────────────────────
        @POST("bookings")
        Call<ApiResponse<BookingResponse>> createBooking(@Body BookingRequest request);

        @POST("bookings/quote")
        Call<ApiResponse<BookingResponse>> getBookingQuote(@Body BookingRequest request);

        @GET("bookings/me")
        Call<ApiResponse<PageResponse<BookingSummary>>> getMyBookings(
                        @Query("page") int page,
                        @Query("size") int size);

        @GET("bookings/{id}")
        Call<ApiResponse<BookingResponse>> getBookingDetail(@Path("id") String id);

        @POST("bookings/check-in")
        Call<ApiResponse<CheckInResponse>> checkIn(@Query("qrCode") String qrCode);

        @GET("bookings/cancel-policy")
        Call<ApiResponse<Map<String, Object>>> getCancelPolicy();

        @POST("bookings/{id}/cancel-request")
        Call<ApiResponse<Void>> cancelRequest(@Path("id") String id);

        @POST("bookings/cancel-confirm")
        Call<ApiResponse<Void>> cancelConfirm(@Query("token") String token, @Query("bookingId") String bookingId);

        // ── Payment ───────────────────────────────────────────────────────────
        @POST("payments")
        Call<ApiResponse<PaymentResponse>> createPayment(@Body PaymentRequest request);

        @GET("payments/booking/{bookingId}")
        Call<ApiResponse<PaymentResponse>> getPaymentStatus(@Path("bookingId") String bookingId);

        @POST("payments/wallet/{bookingId}")
        Call<ApiResponse<PaymentResponse>> payWithWallet(@Path("bookingId") String bookingId);

        // ── Voucher ───────────────────────────────────────────────────────────
        @GET("vouchers/validate")
        Call<ApiResponse<VoucherSummary>> validateVoucher(@Query("code") String code);

        @GET("users/me/vouchers")
        Call<ApiResponse<List<VoucherSummary>>> getMyVouchers();

        @POST("users/me/vouchers/claim")
        Call<ApiResponse<Void>> claimVoucher(@Body ClaimVoucherRequest request);

        @GET("vouchers/active")
        Call<ApiResponse<List<VoucherSyncResponse>>> getActiveVouchers();

        // ── Combo ─────────────────────────────────────────────────────────────
        @GET("combos")
        Call<ApiResponse<List<ComboResponse>>> getCombos();

        // ── Review ────────────────────────────────────────────────────────────
        @GET("reviews")
        Call<ApiResponse<PageResponse<ReviewResponse>>> getReviews(
                        @Query("movieId") String movieId,
                        @Query("page") int page,
                        @Query("size") int size);

        @POST("reviews")
        Call<ApiResponse<ReviewResponse>> createReview(@Body ReviewRequest request);

        // ── GiftCard ─────────────────────────────────────────────────────
        @POST("gift-cards/redeem")
        Call<ApiResponse<GiftCardResponse>> redeemGiftCard(@Body Map<String, String> body);

        @GET("gift-cards/me")
        Call<ApiResponse<PageResponse<GiftCardResponse>>> getMyGiftCards(
                        @Query("page") int page,
                        @Query("size") int size);

        // ── Notification ──────────────────────────────────────────────────────
        @GET("notifications")
        Call<ApiResponse<PageResponse<NotificationResponse>>> getNotifications(
                        @Query("page") int page,
                        @Query("size") int size);

        @GET("notifications/unread-count")
        Call<ApiResponse<UnreadCountResponse>> getUnreadCount();

        @PATCH("notifications/read-all")
        Call<ApiResponse<Void>> markAllAsRead();

        // ── Promotion ─────────────────────────────────────────────────────────
        @GET("promotions/active")
        Call<ApiResponse<List<PromotionResponse>>> getActivePromotions();

        // ── Chatbot ───────────────────────────────────────────────────────────
        @POST("chatbot/chat")
        Call<ApiResponse<Map<String, String>>> chatWithAi(@Body ChatRequest request);

        @POST("chatbot/session/clear")
        Call<ApiResponse<Map<String, String>>> clearChatSession();

        // ── Home Layout ──────────────────────────────────────────────────────────
        @GET("home/featured-movies")
        Call<ApiResponse<List<MovieSummary>>> getFeaturedMovies();

        @GET("home/popup-promotion")
        Call<ApiResponse<PromotionResponse>> getPopupPromotion();

        // ── Staff Dashboard ───────────────────────────────────────────────────
        @GET("staff/dashboard/stats")
        Call<ApiResponse<StaffDashboardStatsResponse>> getStaffDashboardStats();

        @GET("staff/dashboard/upcoming-showtimes")
        Call<ApiResponse<List<UpcomingShowtimeResponse>>> getUpcomingShowtimes();

        @GET("staff/check-in-history")
        Call<ApiResponse<PageResponse<CheckInHistoryItemResponse>>> getCheckInHistory(
                        @Query("page") int page,
                        @Query("size") int size);
}

File: App/app/src/main/java/com/cinema/ticket_booking/network/ApiService.kt
package com.cinema.ticket_booking.network

import com.cinema.ticket_booking.data.model.request.*
import com.cinema.ticket_booking.data.model.response.*
import okhttp3.MultipartBody
import retrofit2.Call
import retrofit2.http.*

interface ApiService {

    // ── Auth ─────────────────────────────────────────────────────────────
    @POST("auth/register")
    fun register(@Body request: RegisterRequest): Call<ApiResponse<AuthResponse>>

    @POST("auth/login")
    fun login(@Body request: LoginRequest): Call<ApiResponse<AuthResponse>>

    @POST("auth/social-login")
    fun socialLogin(@Body request: SocialLoginRequest): Call<ApiResponse<AuthResponse>>

    @POST("auth/refresh")
    fun refreshToken(@Body request: RefreshTokenRequest): Call<ApiResponse<AuthResponse>>

    @POST("auth/logout")
    fun logout(@Body request: RefreshTokenRequest): Call<ApiResponse<Void>>

    @POST("auth/forgot-password")
    fun forgotPassword(@Body request: ForgotPasswordRequest): Call<ApiResponse<Void>>

    @POST("auth/verify-otp")
    fun verifyOtp(@Body body: Map<String, String>): Call<ApiResponse<String>>

    @POST("auth/reset-password")
    fun resetPassword(@Body request: ResetPasswordRequest): Call<ApiResponse<Void>>

    // ── User ─────────────────────────────────────────────────────────────
    @GET("users/me")
    fun getMyProfile(): Call<ApiResponse<UserResponse>>

    @PATCH("users/me")
    fun updateProfile(@Body request: UpdateProfileRequest): Call<ApiResponse<UserResponse>>

    @PATCH("users/me/notifications")
    fun updateNotificationSettings(@Body request: NotificationSettingsRequest): Call<ApiResponse<UserResponse>>

    @Multipart
    @POST("users/me/avatar")
    fun uploadAvatar(@Part file: MultipartBody.Part): Call<ApiResponse<UserResponse>>

    @PUT("users/me/password")
    fun changePassword(@Body body: Map<String, String>): Call<ApiResponse<Void>>

    @PATCH("users/me/fcm-token")
    fun updateFcmToken(@Query("token") token: String): Call<ApiResponse<Void>>

    // ── Movie ─────────────────────────────────────────────────────────────
    @GET("movies")
    fun getMovies(
        @Query("status") status: String,
        @Query("page") page: Int,
        @Query("size") size: Int
    ): Call<ApiResponse<PageResponse<MovieSummary>>>

    @GET("movies/search")
    fun searchMovies(
        @Query("q") query: String,
        @Query("page") page: Int,
        @Query("size") size: Int
    ): Call<ApiResponse<PageResponse<MovieSummary>>>

    @GET("movies/{id}")
    fun getMovieDetail(@Path("id") id: String): Call<ApiResponse<MovieDetail>>

    @GET("movies/{id}/can-review")
    fun canReview(@Path("id") id: String): Call<ApiResponse<CanReviewResponse>>

    @GET("movies/cinema/{cinemaId}")
    fun getNowShowingByCinema(@Path("cinemaId") cinemaId: String): Call<ApiResponse<List<MovieSummary>>>

    @GET("movies/genres")
    fun getAllGenres(): Call<ApiResponse<List<Genre>>>

    // ── Cinema ────────────────────────────────────────────────────────────
    @GET("cinemas")
    fun getCinemas(@Query("city") city: String?): Call<ApiResponse<List<CinemaResponse>>>

    @GET("cinemas/{id}")
    fun getCinemaDetail(@Path("id") id: String): Call<ApiResponse<CinemaResponse>>

    // ── Showtime ──────────────────────────────────────────────────────────
    @GET("showtimes")
    fun getShowtimes(
        @Query("movieId") movieId: String?,
        @Query("cinemaId") cinemaId: String?,
        @Query("date") date: String?
    ): Call<ApiResponse<List<ShowtimeResponse>>>

    @GET("showtimes/{id}/seats")
    fun getSeatMap(@Path("id") showtimeId: String): Call<ApiResponse<SeatMapResponse>>

    // ── Booking ───────────────────────────────────────────────────────────
    @POST("bookings")
    fun createBooking(@Body request: BookingRequest): Call<ApiResponse<BookingResponse>>

    @POST("bookings/quote")
    fun getBookingQuote(@Body request: BookingRequest): Call<ApiResponse<BookingResponse>>

    @GET("bookings/me")
    fun getMyBookings(
        @Query("page") page: Int,
        @Query("size") size: Int
    ): Call<ApiResponse<PageResponse<BookingSummary>>>

    @GET("bookings/{id}")
    fun getBookingDetail(@Path("id") id: String): Call<ApiResponse<BookingResponse>>

    @POST("bookings/check-in")
    fun checkIn(@Query("qrCode") qrCode: String): Call<ApiResponse<CheckInResponse>>

    @GET("bookings/cancel-policy")
    fun getCancelPolicy(): Call<ApiResponse<Map<String, Any>>>

    @POST("bookings/{id}/cancel-request")
    fun cancelRequest(@Path("id") id: String): Call<ApiResponse<Void>>

    @POST("bookings/cancel-confirm")
    fun cancelConfirm(
        @Query("token") token: String,
        @Query("bookingId") bookingId: String
    ): Call<ApiResponse<Void>>

    // ── Payment ───────────────────────────────────────────────────────────
    @POST("payments")
    fun createPayment(@Body request: PaymentRequest): Call<ApiResponse<PaymentResponse>>

    @GET("payments/booking/{bookingId}")
    fun getPaymentStatus(@Path("bookingId") bookingId: String): Call<ApiResponse<PaymentResponse>>

    @POST("payments/wallet/{bookingId}")
    fun payWithWallet(@Path("bookingId") bookingId: String): Call<ApiResponse<PaymentResponse>>

    // ── Voucher ───────────────────────────────────────────────────────────
    @GET("vouchers/validate")
    fun validateVoucher(@Query("code") code: String): Call<ApiResponse<VoucherSummary>>

    @GET("users/me/vouchers")
    fun getMyVouchers(): Call<ApiResponse<List<VoucherSummary>>>

    @POST("users/me/vouchers/claim")
    fun claimVoucher(@Body request: ClaimVoucherRequest): Call<ApiResponse<Void>>

    @GET("vouchers/active")
    fun getActiveVouchers(): Call<ApiResponse<List<VoucherSyncResponse>>>

    // ── Combo ─────────────────────────────────────────────────────────────
    @GET("combos")
    fun getCombos(): Call<ApiResponse<List<ComboResponse>>>

    @GET("reviews")
    fun getReviews(
        @Query("movieId") movieId: String,
        @Query("page") page: Int,
        @Query("size") size: Int,
        @Query("rating") rating: Int?
    ): Call<ApiResponse<PageResponse<ReviewResponse>>>

    @POST("reviews")
    fun createReview(@Body request: ReviewRequest): Call<ApiResponse<ReviewResponse>>

    @PUT("reviews/{id}")
    fun updateReview(
        @Path("id") id: String,
        @Body request: ReviewRequest
    ): Call<ApiResponse<ReviewResponse>>

    // ── GiftCard ─────────────────────────────────────────────────────
    @POST("gift-cards/redeem")
    fun redeemGiftCard(@Body body: Map<String, String>): Call<ApiResponse<GiftCardResponse>>

    @GET("gift-cards/me")
    fun getMyGiftCards(
        @Query("page") page: Int,
        @Query("size") size: Int
    ): Call<ApiResponse<PageResponse<GiftCardResponse>>>

    // ── Notification ──────────────────────────────────────────────────────
    @GET("notifications")
    fun getNotifications(
        @Query("page") page: Int,
        @Query("size") size: Int
    ): Call<ApiResponse<PageResponse<NotificationResponse>>>

    @GET("notifications/unread-count")
    fun getUnreadCount(): Call<ApiResponse<UnreadCountResponse>>

    @PATCH("notifications/read-all")
    fun markAllAsRead(): Call<ApiResponse<Void>>

    @DELETE("notifications/{id}")
    fun deleteNotification(@Path("id") id: String): Call<ApiResponse<Void>>

    // ── Promotion ─────────────────────────────────────────────────────────
    @GET("promotions/active")
    fun getActivePromotions(): Call<ApiResponse<List<PromotionResponse>>>

    // ── Chatbot ───────────────────────────────────────────────────────────
    @POST("chatbot/chat")
    fun chatWithAi(@Body request: ChatRequest): Call<ApiResponse<Map<String, String>>>

    @POST("chatbot/session/clear")
    fun clearChatSession(): Call<ApiResponse<Map<String, String>>>

    // ── Home Layout ──────────────────────────────────────────────────────────
    @GET("home/featured-movies")
    fun getFeaturedMovies(): Call<ApiResponse<List<MovieSummary>>>

    @GET("home/popup-promotion")
    fun getPopupPromotion(): Call<ApiResponse<PromotionResponse>>

    // ── Staff Dashboard ───────────────────────────────────────────────────
    @GET("staff/dashboard/stats")
    fun getStaffDashboardStats(): Call<ApiResponse<StaffDashboardStatsResponse>>

    @GET("staff/dashboard/upcoming-showtimes")
    fun getUpcomingShowtimes(): Call<ApiResponse<List<UpcomingShowtimeResponse>>>

    @GET("staff/check-in-history")
    fun getCheckInHistory(
        @Query("filter") filter: String,
        @Query("page") page: Int,
        @Query("size") size: Int
    ): Call<ApiResponse<PageResponse<CheckInHistoryItemResponse>>>
}
